### PR TITLE
Add source build patch support to VMR manager

### PR DESCRIPTION
**Pull request by premun**
Handles Source Build patches during initialization/synchronization of the VMR

Expects a new property in `source-mappings.json` pointing to where the patches are located in the VMR (later they can be moved):
```
{
    "patchesPath": "src/installer/src/SourceBuild/tarball/patches",
    ...
}
```

Consider we're updating repo `A` from `sha1` to `sha2`, the new synchronization algorithm is as follows:

1. Checkout VMR
2. Clone individual repo A to temp
3. For each file that has a patch in the VMR, copy its `sha1` version from the clone to the VMR (to working tree only)
4. Prepare a diff from `A` for `sha1..sha2` and apply it onto the VMR
5. Apply any Source Build patches associated with the repo
6. Commit all above to the VMR

Resolves https://github.com/dotnet/arcade/issues/10261

### CODE REVIEW

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrInitializer.cs [hunk 1]
```diff
@@ -59,11 +59,13 @@ public async Task InitializeVmr(SourceMapping mapping, string? targetRevision, C
         await ApplyPatch(mapping, patchPath, cancellationToken);
         cancellationToken.ThrowIfCancellationRequested();
         await TagRepo(mapping, commit.Id.Sha);
+        cancellationToken.ThrowIfCancellationRequested();
+        await ApplyVmrPatches(mapping, cancellationToken);
 
-        var description = PrepareCommitMessage(InitializationCommitMessage, mapping, null, commit.Id.Sha, null);
+        var message = PrepareCommitMessage(InitializationCommitMessage, mapping, oldSha: null, commit.Id.Sha, null);
```
Comment (oleksandr-didyk) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrInitializer.cs:65]: Maybe the last param could also have its name specified as `oldSha` now has?

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs [hunk 2]
```diff
@@ -227,7 +242,9 @@ public async Task UpdateVmr(SourceMapping mapping, string? targetRevision, bool
         }
 
         await TagRepo(mapping, toRevision);
-
+        cancellationToken.ThrowIfCancellationRequested();
+        await ApplyVmrPatches(mapping, cancellationToken);
+        cancellationToken.ThrowIfCancellationRequested();
```
Comment (oleksandr-didyk) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs:247]: Should this line be duplicated before and after the async task?

**Review by oleksandr-didyk — APPROVED** [review]

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs (hunk 2 above)
Comment (premun) [reply src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs:247]: Yes, we want to check after each operation whether we're not being cancelled

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrInitializer.cs (hunk 1 above)
Comment (premun) [reply src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrInitializer.cs]: I've made the nullable parameters optional to make this better

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs (hunk 2 above)
Comment (oleksandr-didyk) [reply src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs:247]: My bad, I did not notice the async task on line 244

**Review by oleksandr-didyk — APPROVED** [review]

**Pull request merged**